ompi/mca/coll/han/coll_han_alltoall.c

```diff
@@ -0,0 +1,496 @@
+    }
+
+    int rc;
+    size_t sndsize;
+    MPI_Aint sextent, rextent, lb;
+    char *send_bounce;
+    char *gather_bounce;
+
+    ompi_communicator_t *low_comm = han_module->sub_comm[INTRA_NODE];
+    ompi_communicator_t *up_comm = han_module->sub_comm[INTER_NODE];
+    ompi_request_t **inter_recv_reqs;
+    ompi_request_t **intra_gather_reqs;
+    ompi_request_t **inter_send_reqs;
+
+    int fanout = mca_coll_han_component.han_alltoall_pstages;
+    if (!fanout) {
+        fanout = 1;
+    }
+    int subdiv_factor = 1;
+
+    OPAL_OUTPUT_VERBOSE((90, mca_coll_han_component.han_output, "Starting mca_coll_han_alltoall_using_allgather\n"));
+
+    int w_rank = ompi_comm_rank(comm);
+    int w_size = ompi_comm_size(comm);
+
+
+    /* information about sub-communicators */
+    int low_rank = ompi_comm_rank(low_comm);
+    int low_size = ompi_comm_size(low_comm);
+    int up_size = ompi_comm_size(up_comm);
+
+    while (low_size % subdiv_factor != 0) {
+        subdiv_factor--;
+        assert(subdiv_factor > 0);
+    }
+    int up_rank = w_rank / low_size;
+    assert( w_rank % low_size == low_rank );
+
+    rc = ompi_datatype_get_extent( sdtype, &lb, &sextent);
+    rc = ompi_datatype_get_extent( rdtype, &lb, &rextent);
+    opal_datatype_type_size( &sdtype->super, &sndsize);
+
+    MPI_Aint gather_bytes = w_size * low_size * rextent * rcount;
+    MPI_Aint send_bytes_per_fan = low_size * rextent * rcount;
+
+    int nreq_intra_per_round = low_size;
+    gather_bounce = malloc(gather_bytes);
+    send_bounce = malloc(send_bytes_per_fan * fanout);
+    intra_gather_reqs = malloc(sizeof(*intra_gather_reqs)*nreq_intra_per_round*fanout);
+    inter_send_reqs = malloc(sizeof(*inter_send_reqs) * fanout);
+    inter_recv_reqs = malloc(sizeof(*inter_recv_reqs) * up_size );
+
+    int nrounds = up_size;
+    /* pre-post all our receives.  We will be ready to receive all data regardless of fan-out.
+       (This is not an in-place algorithm)*/
+    for (int jround=0; jround<nrounds; jround++) {
+            int up_partner = ring_partner(up_rank, jround, up_size);
+            int first_remote_rank = up_partner*low_size;
+            /* pre-post the receive for remote.  Receive directly into application buffer */
+            char *recv_chunk = ((char*)rbuf) + rextent*rcount*first_remote_rank;
+            MCA_PML_CALL(irecv
+                        (recv_chunk, rcount*low_size, rdtype, first_remote_rank+low_rank,
+                        MCA_COLL_BASE_TAG_ALLTOALL,
+                        comm, &inter_recv_reqs[jround]));
+    }
+
+    rc = low_comm->c_coll->coll_allgather(sbuf, scount*w_size, sdtype, gather_bounce, scount*w_size, sdtype, low_comm, low_comm->c_coll->coll_allgather_module);
+
+    if (rc != 0) {
+        OPAL_OUTPUT_VERBOSE((40, mca_coll_han_component.han_output,
+        "Allgather failed with %d\n",rc));
+        goto error;
+    }
+
+    for (int jloop=0; jloop<nrounds+fanout; jloop++) {
```

Comment:
How much performance benefit by overlapping the intranode and internode communication?

Reply:
Not as much as I'd hoped.  My goal wasn't so much to overlap internode and intranode, but rather pipeline internode communication.  You commented on the "using allgather" variant which doesn't perform well in large message sizes.  My observation was that for large message sizes in the smsc variant having two or perhaps three rounds in-flight at once saved me a little bit of time.  However there is an additional penalty I had to pay.  you'll notice for fanout=1 in the smsc variant I don't allocate a send_bounce buffer at all, and just re-use the application buffer.  This sidesteps the need for transport to perform ibv_mr_reg on a temporary buffer, which might change every call.  By re-using the application buffer the mr_reg cache happily re-uses previous registrations everytime and this typically outweighs the fanout improvements.

If we had an ability to re-use a buffer pool of registered memory, the fanout might be more helpful.